### PR TITLE
Update: Extract glob and filesystem logic from cli-engine (fixes #4305)

### LINKED ISSUE #4305: File-list building functionality could be extracted into a utility
Currently, `cli-engine.js` contains all of the logic for converting a set of patterns (globs, filenames, and paths) into a list of files to process/lint, and much of this is within the `executeOnFiles` method.  This logic could be broken out into a separate utility, so that it could be used in other areas, for example autoconfig (#3567).

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/27847586-file-list-building-functionality-could-be-extracted-into-a-utility?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by IanVS**
This will also make some of the tricky glob functions more testable.

### CODE REVIEW

**Comment by IanVS** [comment]
@BYK IT would be great if you could think of any other edge cases which should be tested here.  The tests should also probably be run on Windows, since we seem to have problems with globs there sometimes.

File: lib/util/glob-util.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        }
+
+        return newPath.replace(/\\/g, "/").replace(/^\.\//, "");
+    };
+}
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
+
```
Comment (nzakas) [inline lib/util/glob-util.js]: Please use the form

```js
module.exports = {

};
```

This is what we use throughout the code base.

Comment (IanVS) [inline lib/util/glob-util.js]: Too bad there's nothing out there which could help us enforce consistent style in our code base. :trollface: 

I've updated to use the desired form.

I had noticed the `exports.` form in `rules.js`, so I have now fixed it there and in one other place as well.

Comment (ilyavolodin) [inline lib/util/glob-util.js]: Might be a good idea to create a rule for this. Although this would be hard to do.

Comment (IanVS) [inline lib/util/glob-util.js]: The opposite of this was proposed in: https://github.com/eslint/eslint/issues/3637 and determined not to be a good fit for core.

File: lib/cli-engine.js [hunk 2]
```diff
@@ -550,18 +505,8 @@ CLIEngine.prototype = {
         loadedPlugins[pluginNameWithoutPrefix] = pluginobject;
     },
 
-    /**
-     * Resolves the patterns passed into executeOnFiles() into glob-based patterns
-     * for easier handling.
-     * @param {string[]} patterns The file patterns passed on the command line.
-     * @returns {string[]} The equivalent glob patterns.
-     */
     resolveFileGlobPatterns: function(patterns) {
```
Comment (ilyavolodin) [inline lib/cli-engine.js]: Don't think we need this anymore. Just change all of the invocations to `globUtil`

File: lib/cli-engine.js [hunk 3]
```diff
@@ -619,22 +557,16 @@ CLIEngine.prototype = {
...
          */
         function executeOnFile(filename) {
-            var shouldIgnore = ignoredPaths.contains(filename);
-            if (shouldIgnore) {
-                return;
-            }
+            var hashOfConfig;
 
-            filename = fs.realpathSync(filename);
             if (processed[filename]) {
```
Comment (ilyavolodin) [inline lib/cli-engine.js]: I'm not sure this is necessary anymore, since you are resolving all directories to files up front anyways.

File: lib/cli-engine.js (hunk 2 above)
Comment (IanVS) [inline lib/cli-engine.js]: It was a public API. Wouldn't it require a breaking change to remove?

Comment (ilyavolodin) [inline lib/cli-engine.js]: Ahh.. sorry, didn't notice that. In that case you should leave jsdoc for it.
